# 两自由度欠驱动系统例子

两自由度的线性系统，动力学方程为：

$$
M \begin{pmatrix}
    \ddot{x}_{1} \\ \ddot{x}_{2}
\end{pmatrix} + C \begin{pmatrix}
    \dot{x}_{1} \\ \dot{x}_{2}
\end{pmatrix} + K \begin{pmatrix}
    x_{1} \\ x_{2}
\end{pmatrix} = \begin{bmatrix}
    u \\ 0
\end{bmatrix}
$$

其中，

$$
M = \begin{bmatrix}
    m_{1} & 0 \\
    0 & m_{2}
\end{bmatrix},
C = \begin{bmatrix}
    c_{1} + c_{2} & - c_{2} \\
    - c_{2} & c_{1} + c_{2}
\end{bmatrix},
K = \begin{bmatrix}
    k_{1} + k_{2} & - k_{2} \\
    - k_{2} & k_{1} + k_{2}
\end{bmatrix}.
$$

参考《振动力学》P247例子，物理参数设定为

$$
\begin{aligned}
    m_{1} = 1(kg), & m_{2} = 1.5(kg), \\
    c_{1} = 0.6284 (N \cdot s/m), & c_{2} = 0.0628 (N \cdot s/m), \\
    k_{1} = 987 (N/m), & k_{2} = 217 (N/m).
\end{aligned}
$$


In [4]:
import numpy as np
from IPython.display import display, Latex

m1 = 1
m2 = 1.5
c1 = 0.6284
c2 = 0.0628
k1 = 987
k2 = 217

M = np.array([[m1, 0], [0, m2]])
C = np.array([[c1 + c2, -c2], [-c2, c1 + c2]])
K = np.array([[k1 + k2, -k2], [-k2, k1 + k2]])

print("M =", M)
print("C =", C)
print("K =", K)


M = [[1.  0. ]
 [0.  1.5]]
C = [[ 0.6912 -0.0628]
 [-0.0628  0.6912]]
K = [[1204 -217]
 [-217 1204]]


计算得到质量阵、阻尼阵、刚度阵分别为

$$
M = \begin{bmatrix}
    1 & 0 \\
    0 & 1.5
\end{bmatrix},
C = \begin{bmatrix}
    0.6912 & -0.0628 \\
    -0.0628 & 0.6912
\end{bmatrix},
K = \begin{bmatrix}
    1204 & -217 \\
    -217 & 1204
\end{bmatrix}.
$$


In [9]:
SystemMatrix = np.concatenate(
    (np.concatenate((np.zeros((2, 2)), np.identity(2)), axis=1),
     np.concatenate((-np.linalg.inv(M) @ K, -np.linalg.inv(M) @ C), axis=1)),
    axis=0)
InputMatrix = np.array([[0], [0], [1], [0]])

display(Latex("$\mathcal{A} = $"))
print(SystemMatrix)


<IPython.core.display.Latex object>

[[ 0.00000000e+00  0.00000000e+00  1.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]
 [-1.20400000e+03  2.17000000e+02 -6.91200000e-01  6.28000000e-02]
 [ 1.44666667e+02 -8.02666667e+02  4.18666667e-02 -4.60800000e-01]]


可写出状态空间方程：

$$
\begin{aligned}
    \frac{d}{d t} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix} &= \begin{bmatrix}
        \mathbf{0} & I \\
        - M^{-1} K & - M^{-1} C
    \end{bmatrix} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix} + \begin{pmatrix}
        0 \\ 0 \\ 1 \\0
    \end{pmatrix} u \\
    &= \mathcal{A} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix} + \begin{pmatrix}
        0 \\ 0 \\ 1 \\0
    \end{pmatrix} u
\end{aligned}
$$

其中，

$$
\begin{aligned}
\mathcal{A} &= \begin{bmatrix}
        \mathbf{0} & I \\
        - M^{-1} K & - M^{-1} C
    \end{bmatrix} \\
    &= \begin{bmatrix}
        0 & 0 & 1 & 0 \\
        0 & 0 & 0 & 1 \\
        -1204 & 217 & -0.6912 & 0.0628 \\
        144.667 & -802.667 & 0.0418667 & -0.4608
    \end{bmatrix}
\end{aligned}
$$


定义CLF函数

$$
V = \frac{1}{2} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix}^{\top} H \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix}
    = \frac{1}{2} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix}^{\top} \begin{bmatrix}
        1 & 0 & 0 & 0 \\
        0 & 0.5 & 0 & 0 \\
        0 & 0 & 0.01 & 0 \\
        0 & 0 & 0 & 0.01
    \end{bmatrix} \begin{pmatrix}
        x_{1} \\ x_{2} \\ \dot{x}_{1} \\ \dot{x}_{2}
    \end{pmatrix}
$$

要求找到增益的取值范围，使得CLF函数满足如下指数稳定性

$$
\frac{d V}{d t} + \lambda V < 0
$$

在本例中，取$\lambda = 0.8$

In [6]:
H = np.diag(np.array([1, 0.5, 0.01, 0.01]))
print("H =", H)


H = [[1.   0.   0.   0.  ]
 [0.   0.5  0.   0.  ]
 [0.   0.   0.01 0.  ]
 [0.   0.   0.   0.01]]


验证系统的可控性


In [18]:
np.linalg.matrix_rank(
    np.concatenate((InputMatrix, SystemMatrix @ InputMatrix,
                    SystemMatrix @ SystemMatrix @ InputMatrix,
                    SystemMatrix @ SystemMatrix @ SystemMatrix @ InputMatrix),
                   axis=1)) == SystemMatrix.shape[0]


True

若考虑数字采样和时滞，作如下假设。假设系统的采样周期为$\tau = 0.01s$，当$t = \lfloor \frac{t}{\tau} \rfloor \tau + t_{0}= N \tau + t_{0}$时

$$
u = k_{p} x_{1}(N \tau) + k_{d} \dot{x}_{2}((N - 1) \tau)
$$
